### DISCUSSION TITLE
Execute more than one query per page

### DISCUSSION THREAD

**Discussion by lucashfreitas** [gatsbyjs/gatsby#32217] [Help]
Hello,

My graphql queries are executing the limit/quota from my backend CMS, both on the query size in bytes and complexity. I tried to minify the query string but I am still having the issue. 

Is there any way to split/execute more than a single query per page? I thought about other alternatives as moving the queries to `gatsby-node` and passing the data via context or even changing it to static queries and executing two different queries but it would require a massive amount of work to restructure all my codebase/pages.

Does anyone know if that is possible or what could be another alternative?

Thanks!

**Comment by LekoArts** [comment]
You would need to adapt your source plugin, not the page queries itself. Because the requested data will be the same independent of where you put it.

Is this a custom source plugin or a CMS with a plugin?

**Comment by lucashfreitas** [comment]
Hello @LekoArts , I am using the `gatsby-source-graphql` plugin and contentful CMS. I have a lot of different data/fields to query for each page and it's exceeding contentful limits, so I want to split the queries on every page but not too sure if this is possible using this plugin or gatsby. It seems that gatsby can only execute a single graphql query per page, right?

**Reply by KyleAMathews** [reply]
why not use gatsby-source-contentful?

**Reply by lucashfreitas** [reply]
@KyleAMathews the websites I have been working in the past are medium to big ( 1000+) pages and gatsby-source-contentful has been considerately slower than `gatsby-source-graphql` and we are having a hard time to set up previews without gatsby cloud, so since contentful launched graphql api I have been using it. 

At this stage would be a bit of work to change it to `gatsby-source-contentful` and it's too risky to do that as using `gatsby-source-graphql` the website is taking around 5 minutes to build, so I imagine it will be even more with gatsby-source-contentful. I would also have to modify all queries.

I am also trying to set up `preview` at the moment but unfortunately things aren't working out of the box because of contentful graphql api and lack of incremental build (I am using gatsby v2 here - will try to make it work before doing the migration to V3 as I might have to test the whole website and/or fix incompatibilities). Our biggest problem since we start developing medium/big websites in gatsby has been the buildtime + preview feature and I am trying to find a definitive solution that does not requires to add a new cloud service as Gatsby cloud. 

I will keep investigating/trying to find some solution and will post it here once I have some news.

Sorry for throwing another question here:

Is there any documentation about `graphql` function used on `createPages` function? Is this the same as global graphql function? I am creating some pages dynamically in `gatsby-node.js` and would like to pass variables to the graphql query, but it seems that the function accepts only a single parameter which is the graphql query string: 

```
exports.createPages = (params) => {
  const { graphql } = params; 
   
  graphql(`query .....`, VARIABLES_HERE?`);


```

Thank you very much :)

**Reply by lucashfreitas** [reply]
@KyleAMathews another very strong reason why I have not used the `gatsby-souce-contentful` is that it automatically infers the type, so let's suppose we have a field in the backend that can have a title or not. If we don't have any entry with a title the plugin won't be smart to infer that and create the `nodefield` and it will return an error - this has been a big issue in the past. Would be super nice if the plugin was smart enough to read the content types and based on it generated the Nodes fields. 

I know we can customize schema do get around this but would be a bit overkilling to do that manually for every content type, specially for big websites. ` gatsby-source-graphql` is nice and clean and also has a plugin that can generate types which is a massive benefit if you're using typescript.

I am not 100% sure why the source plugins uses this "automatic" way of inferring data types if the CMSs always have the data structure in their end and it's the source of truth. The problem I have been facing for years and have seen many gatsby users facing the same is: 

> let's supose we have 10 entries that has a **title** property but the user didn't fill any of them, if you try to query this field using the the majority of source-plugins (including gatsby-source-contentful), it will return an error saying that the field doesn't exists. Considering that the structure,field types can easily be retrieved from the CMS and it is the single source of truuth I am not 100% sure why the plugins try to infer graphql schema types instead of building the nodes based on the content type according to each CMS.

**Reply by lucashfreitas** [reply]
I will raise another discussion around that.

**Reply by KyleAMathews** [reply]
Incremental builds with source-contentful on cloud generally take seconds btw.

**Reply by lucashfreitas** [reply]
Yeap - I have enabled preview in gatsby cloud using gatsby-source-graphql and is taking 2 minutes :(
